Search path fixes, v2

abbradar opened the pull request:
###### Things done
- [ ] Tested using sandboxing (`nix-build --option build-use-chroot true` or [nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This is continuation of my attempts at having outputs correctly used tree-wise. We now have a more specific function `makeSearchPathOutput`, defined via `getOutput` -- a new function for getting output from a package with necessary checks and fallbacks. This function is very useful on its own, so I've run a few regexps tree-wise to use it instead of ad-hoc `a.needed or a.out or a`. What I _don't_ like is that as of now ideally this function should be used anywhere to abstract from splitted/unsplitted packages, and that would be a slowdown and also clunky-looking. However it's better than ad-hoc fixes anyway and this is a price for multiple outputs -- a little one for its benefits ~_^.

cc @vcunat @edolstra

abbradar committed:
Revert "tryAttrs: init function"

This reverts commit 338340f993563551d8cb45941da987408abef65f.

abbradar committed:
getOutput function: init

abbradar committed:
makeSearchPathOutputs: refactor to makeSearchPathOutput

abbradar committed:
make*Path: implement via makeSearchPathOutput

abbradar committed:
add get* helper functions and mass-replace manual outputs search with them

7c6f434c commented:
Looks nice

abbradar commented:
I had an idea right now to instead define `out`, `lib`, `dev` and `bin` everywhere, aliasing them as needed for packages that are not splitted. What do you think?

jcumming commented:
This is just the tool I was looking for to fix the torbrowser derivation..

abbradar commented:
Notice that you can do it already -- just use `makeLibraryPath`, `makeBinPath` or `makeSearchPathOutputs "needed/path" ["neededOutput"]` (the latter will be replaced in this PR -- ping me to include an update).

cstrahan commented:
This looks pretty great. I'd like to see this and #14766 merged.

abbradar commented:
This looks OK to merge for me, but first I'd like to gather feedback on an idea briefly described in https://github.com/NixOS/nixpkgs/pull/14694#issuecomment-210045152. In more details:

Right now we are slowly moving through splitting more and more packages. Also, we have a lot of packages that refer to other packages' outputs explicitly for various reasons, like `FOO_PATH = "${foo}/bin/foo`. This would break when `foo` is split (well, in this case it would not break after #14766 but then it would break `${foo}/include`). So, to simplify splitting packages in future new code would likely be written to abstract over split outputs, as it is already written now; for example, `FOO_PATH = "${foo.dev or foo.out or foo}/include"`. Of course writing this manually is too cumbersome, so this PR introduces `lib.getOutput` which covers all the details you need to consider (i.e. `outputUnspecified`, checking `out` and so on). However, writing `lib.getOutput "dev" foo` everywhere is cumbersome too, and so (arguably) is `lib.getDev foo` (or `stdenv.lib.getDev foo`!).

What I propose instead is to add several most popular attributes to all packages _by default_. They would point to needed outputs in case the package was really split, and would be an alias if it wasn't. I think the most used ones are `lib`, `dev` and `bin`. Of course there are more, but those are 95% of cases (subjectively -- I don't have numbers) of all "explicit usage" (i.e. not in `environmentPackages`, `buildInputs` or other places which can implement their own choosing logic). This would simplify described problem to just `foo.dev` without pulling `lib` and using a function.

What do you think? I'm willing to try to implement that if people are interested.

cstrahan commented:
@abbradar I don't have any strong opinions. On the surface, that sounds nice though.

/cc @edolstra -- I figure you might have opinions wrt https://github.com/NixOS/nixpkgs/pull/14694#issuecomment-211098084.

vcunat commented on lib/attrsets.nix:
```diff
@@ -438,23 +438,25 @@ rec {
+       else
+         pkg.out or pkg
+      )
+    else pkg;
```
Why such a complex code? I'd write
```nix
if pkg.outputUnspecified or false
  then pkg.${output} or pkg.out or pkg
  else pkg
```

abbradar replied:
Oh, that's because I'm constantly forgetting that I can do `pkg.${foo}` in Nix!

vcunat commented on pkgs/stdenv/generic/default.nix:
```diff
@@ -223,7 +223,7 @@ let
         meta = { }
             # If the packager hasn't specified `outputsToInstall`, choose a default,
-            # namely `p.bin or p.out or p`;
+            # namely `lib.getBin p`;
```
Why not really use `getBin` a few lines below this?

abbradar replied:
You mean [there](https://github.com/abbradar/nixpkgs/blob/search-path-fixes/pkgs/stdenv/generic/default.nix#L235)? That's because the logic is a bit different -- if not `bin` nor `out` is found it iterates by `outs` instead of just getting the default (unspecified output).

vcunat replied:
- I believe the logic is the same. The default output is the first one in the `outputs` list.
- Especially if the logic wasn't the same, the comment should correspond to the implementation. Otherwise it's confusing.

abbradar replied:
Hm, what do you mean by "default output", `pkg`? If so I cannot see how can it be chosen there, because we iterate over some list and try each _attribute_ without ever resorting to the attribute set itself. If nothing was found, we'd just return `[null]`.

vcunat replied:
The code
```nix
lib.findFirst hasOutput null (["bin" "out"] ++ outs)
```
first tries "bin" output then "out" and then the first output (`outs` at this point is the list of output names).

abbradar replied:
Hm, indeed, but `getBin` would try `pkg.bin`, `pkg.out` and then just `pkg`. There, however, we'll try `pkg.bin`, `pkg.out`, `pkg.{foo,bar}` (where `outs = [ "foo" "bar" ]`) and then return `null` (without trying `pkg`).

vcunat replied:
When `outs = [ "foo" ...]`, `pkg == pkg.foo` (except for `.outputUnspecified`), and consequently `hasOutput "foo"` will succeed and `findFirst` will never proceed further.

This whole discussion only strengthens my conviction that the code should be rewritten to use `getBin`...

abbradar replied:
A-ha, I presume that this constraint is because of the surrounding code? I judged it only by this line alone, sorry.

Then we can just replace this whole line by `[getBin out]`, right? (just to confirm, because from just looking at the line this changes the logic and I don't know the context good enough)

vcunat replied:
Yes, the semantics is `[(getBin pkg)]`. Thinking of it again, I can't see how to get a reasonable `pkg` argument on that line to pass it, so it might actually be best to leave the PR as it is now :-)

vcunat replied:
Actually, no &ndash; there's the *name* of the output and not the output itself. Now I can't see anything better than the current state of the PR (around this point).

abbradar replied:
Maybe we want to update the comment in some way to indicate what's going on here?

vcunat replied:
Perhaps:
``` nix
# namely the name of `p.bin or p.out or p`
```

abbradar replied:
Done! Thanks for explanations.

vcunat replied:
:-) "namely the name" sounds really awful, but it should be clear at least...

abbradar replied:
<bikeshedding mode on>I've replaced "namely" with "which is", looks better.

vcunat commented:
> I had an idea right now to instead define out, lib, dev and bin everywhere, aliasing them as needed for packages that are not splitted.

I had thought of that briefly before. It seems the best way I know of to go in long term. To do that properly, these would have to react to vars like `outputDev`, implying that the [process of selecting defaults](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/setup-hooks/multiple-outputs.sh#L31) should be done during evaluation and not during build. The complexity of that change is what had dissuaded me from doing it straight away. (The branch was running for years even without adding such changes.)

vcunat commented:
This PR looks good to me. (Mainly the general stuff; I only skimmed the package-specific changes.) I suppose we should go with it (or most of it) until someone implements that bigger plan outlined in my previous comment.

abbradar commented:
Hm, I think that we indeed can merge it as is -- `get{Dev,Lib,Bin}` is quite regex-able functions, so when (if) we implement the plan we can just go over the code, replace them where found and drop `get???` from the library.

Rebased and updated as per @vcunat's comments (and replaced all the new `makeLibraryPathOutputs` usages).

abbradar committed:
stdenv: clarify how `outputsToInstall` is chosen

See https://github.com/NixOS/nixpkgs/pull/14694/files#r60013871

edolstra commented on nixos/modules/system/boot/loader/grub/grub.nix:
```diff
@@ -58,7 +58,7 @@ let
         pkgs.coreutils pkgs.gnused pkgs.gnugrep pkgs.findutils pkgs.diffutils pkgs.btrfs-progs
         pkgs.utillinux ] ++ (if cfg.efiSupport && (cfg.version == 2) then [pkgs.efibootmgr ] else [])
-      )) + ":" + (makeSearchPathOutputs "sbin" ["bin"] [
+      )) + ":" + (makeSearchPathOutput "bin" "sbin" [
```
Adding `sbin` to search paths is obsolete, since we automatically move `sbin/*` to `bin/`.

abbradar replied:
Hm, I saw several places in code where `sbin` was accounted for. Do we always move it, or are there exceptions?

vcunat replied:
I don't know of a possible reason not to move it. The occurrences are most likely just legacy.

vcunat replied:
Note that `dontMoveSbin` is unused in nixpkgs.

abbradar replied:
I thought of looking through all our `sbin` uses and replace them, but we have too much of them: `grep -r 'sbin' | wc -l` shows 835 occurrences. I think removing them is out of scope of this PR, although I can remove those that I touched if desired.

vcunat replied:
It would be more fitting for a separate mass-replace commit. I'd still probably leave the automatic sbin -> bin symlinks.

abbradar replied:
Yeah, I thought of just removing things like mentions in `PATH`.

abbradar committed:
octoprint: use makeSearchPathOutput

abbradar committed:
cmake2_8: use get(Output) functions

abbradar committed:
atomEnv: use makeSearchPathOutput

abbradar commented:
I'm getting a mass rebuild now which shouldn't be the case. How do we usually debug this? `.drv`s are not very nice to `diff`...

abbradar commented:
Ouch, it accidentally uncovered a bug in gcc wrapper D:
Specifically, old code in `pkgs/build-support/cc-wrapper/default.nix`:
```
  binutils_bin = if nativeTools then "" else binutils.bin or binutils;
```
This incorrectly resolved to `binutils.dev` (because `.bin` doesn't exist). It resolves to `binutils.out` now.

abbradar committed:
gcc-wrapper-old: fix binutils and coreutils' paths

abbradar commented:
I'd say problems like these is why we need to encourage everyone to specify exactly what they mean instead just a package and get the "get several outputs to all derivations as aliases" plan finished sooner (so that keystroke laziness would be less an issue). `${foo => include/foo.h}` would help, too (although not in this case).

vcunat commented:
>  `.drv`s are not very nice to diff...

Just a few hours ago I used `kdiff3` on derivations and it handled them quite well, wrapping and highlighting the differing stuff. (It helped me to find the root of #14965.)

abbradar commented:
In the end I used `meld` because that was the first thing showed up in DDG. I guess I need to find how to do this properly in Emacs...

So, what else should be done before merging this? I'm building `hello` now, just in case.

vcunat commented:
I suppose we can stage it. The risk of mass breakage seems low enough.

abbradar commented:
`hello` built successfully, so it should be okay I think. We have more things that we might want to stage simultaneously:
https://github.com/NixOS/nixpkgs/pull/14909 (simple enough)
https://github.com/NixOS/nixpkgs/pull/14907 (sounds ready, and given that my PR does a little in terms of actual build changes can be the primary test subject)

vcunat commented:
Sounds fine.

abbradar commented:
Closed by https://github.com/NixOS/nixpkgs/commit/5f19542581ed03b5059ffb78113b99f190cc6b29
